pbft/pbft.cpp

```diff
@@ -978,6 +992,533 @@ pbft::get_high_water_mark()
+    for (int i{1}; i < newview_msg.pre_prepare_messages_size(); ++i)
+    {
+        uint64_t current_sequence{extract_sequence(newview_msg.pre_prepare_messages(i))};
+        if (last_sequence + 1 != current_sequence)
+        {
+            return false;
+        }
+        last_sequence = current_sequence;
+    }
+    return last_sequence==latest_sequence;
+}
+
+// prepared_proofs is not a member of a new_view message. You need to
+// iterate over all the prepared_proofs in all the viewchange messages
+// in the new_view.
+uint64_t
+pbft::last_sequence_in_newview_preprepare_messages(const pbft_msg &newview)
+{
+    uint64_t last_sequence{0};
+    for(int i{0}; i < newview.viewchange_messages_size(); ++i)
+    {
+        bzn_envelope viewchange_envelope = newview.viewchange_messages(i);
+        pbft_msg viewchange;
+        viewchange.ParseFromString(viewchange_envelope.pbft());
+
+        for(int j{0}; j < viewchange.prepared_proofs_size(); ++j)
+        {
+            prepared_proof proof = viewchange.prepared_proofs(j);
+            for(int k{0}; k < proof.prepare_size(); ++k)
+            {
+                bzn_envelope prepare = proof.prepare(k);
+                pbft_msg idfk;
+                idfk.ParseFromString(prepare.pbft());
+                last_sequence = std::max(last_sequence, idfk.sequence());
+            }
+        }
+    }
+    return last_sequence;
+}
+
+bool
+pbft::is_valid_newview_message(const pbft_msg& thiers, const bzn_envelope& original_thiers) const
+{
+    // - does it contain 2f+1 viewchange messages
+    if (static_cast<size_t>(thiers.viewchange_messages_size()) < 2 * this->max_faulty_nodes() + 1)
+    {
+        return false;
+    }
+
+    std::set<std::string> viewchange_senders;
+    pbft_msg viewchange_msg;
+    for (int i{0};i < thiers.viewchange_messages_size();++i)
+    {
+        bzn_envelope original_msg{thiers.viewchange_messages(i)};
+        // - are each of those viewchange messages valid?
+        if (!viewchange_msg.ParseFromString(original_msg.pbft()) ||
+            !this->is_valid_viewchange_message(viewchange_msg, original_msg))
+        {
+            LOG(error) << "is_valid_newview_message - new view message contains invalid viewchange message";
+            return false;
+        }
+        viewchange_senders.insert(original_msg.sender());
+    }
+
+    if (viewchange_senders.size() != size_t(thiers.viewchange_messages_size()))
+    {
+        LOG (error) << "is_valid_newview_message - unexpected viewchange message count";
+        return false;
+    }
+
+    std::map<uuid_t,bzn_envelope> viewchange_envelopes_from_senders;
+
+    for (int i{0};i < thiers.viewchange_messages_size();++i)
+    {
+        auto viewchange_env = thiers.viewchange_messages(i);
+        viewchange_envelopes_from_senders[viewchange_env.sender()] = viewchange_env;
+    }
+
+    pbft_msg ours = this->build_newview( thiers.view(), viewchange_envelopes_from_senders);
```

Comment:
remove space after (

Reply:
Done, also fixed another instance of that...